src/libsbn.hpp

```diff
@@ -138,6 +138,31 @@ struct SBNInstance {
     return results;
   }
 
+  std::vector<std::vector<double>> BranchGradients() {
+    if (beagle_instances_.size() == 0) {
+      std::cerr << "Please add some BEAGLE instances that can be used for "
+                   "computation.\n";
+      abort();
+    }
+    std::vector<std::vector<double>> results(tree_collection_->TreeCount());
+    std::queue<beagle::BeagleInstance> instance_queue;
+    for (auto instance : beagle_instances_) {
+      instance_queue.push(instance);
+    }
+    std::queue<size_t> tree_number_queue;
+    for (size_t i = 0; i < tree_collection_->TreeCount(); i++) {
+      tree_number_queue.push(i);
+    }
```

Comment:
I guess we can put the above into some separate function to avoid repetition here.